plugins/core/IO/qLASIO/src/LasIOFilter.cpp

```diff
@@ -0,0 +1,664 @@
+//#                                                                        #
+//#                CLOUDCOMPARE PLUGIN: LAS-IO Plugin                      #
+//#                                                                        #
+//#  This program is free software; you can redistribute it and/or modify  #
+//#  it under the terms of the GNU General Public License as published by  #
+//#  the Free Software Foundation; version 2 of the License.               #
+//#                                                                        #
+//#  This program is distributed in the hope that it will be useful,       #
+//#  but WITHOUT ANY WARRANTY; without even the implied warranty of        #
+//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         #
+//#  GNU General Public License for more details.                          #
+//#                                                                        #
+//#                   COPYRIGHT: Thomas Montaigu                           #
+//#                                                                        #
+//##########################################################################
+
+#include "LasIOFilter.h"
+#include "LasOpenDialog.h"
+#include "LasSaveDialog.h"
+#include "LasSavedInfo.h"
+#include "LasScalarFieldLoader.h"
+#include "LasScalarFieldSaver.h"
+#include "LasWaveformLoader.h"
+#include "LasWaveformSaver.h"
+
+#include <GenericProgressCallback.h>
+#include <ccPointCloud.h>
+#include <ccProgressDialog.h>
+#include <ccScalarField.h>
+
+#include <QDate>
+#include <QElapsedTimer>
+#include <QFileInfo>
+
+#include <laszip/laszip_api.h>
+
+#include <ccColorScalesManager.h>
+#include <memory>
+#include <numeric>
+#include <utility>
+
+constexpr const char *LAS_METADATA_INFO_KEY = "LAS.savedInfo";
+
+static CCVector3d GetGlobalShift(FileIOFilter::LoadParameters &parameters,
+                                 bool &preserveCoordinateShift,
+                                 const CCVector3d &lasOffset,
+                                 const CCVector3d &firstPoint)
+{
+    ccGlobalShiftManager::Mode csModeBackup = parameters.shiftHandlingMode;
+    CCVector3d shift{};
+    bool useLasOffset = false;
+
+    // set the lasOffset as default if none was provided
+    if (lasOffset.norm2() != 0 &&
+        (!parameters._coordinatesShiftEnabled || !*parameters._coordinatesShiftEnabled))
+    {
+        if (csModeBackup !=
+            ccGlobalShiftManager::NO_DIALOG) // No dialog, practically means that we don't want any shift!
+        {
+            useLasOffset = true;
+            shift = -lasOffset;
+            if (csModeBackup != ccGlobalShiftManager::NO_DIALOG_AUTO_SHIFT)
+            {
+                parameters.shiftHandlingMode = ccGlobalShiftManager::ALWAYS_DISPLAY_DIALOG;
+            }
+        }
+    }
+
+    FileIOFilter::HandleGlobalShift(firstPoint, shift, preserveCoordinateShift, parameters, useLasOffset);
```

Comment:
Too be on the safe side, if HandleGlobalShift returns false, you should force 'preserveCoordinateShift' to false, and 'shift' to (0, 0, 0).